x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/CategorizeTextAggregator.java

```diff
@@ -0,0 +1,208 @@
+
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.ObjectArray;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.CardinalityUpperBound;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.bucket.DeferableBucketAggregator;
+import org.elasticsearch.search.aggregations.bucket.terms.LongKeyedBucketOrds;
+import org.elasticsearch.search.aggregations.bucket.terms.TermsAggregator;
+import org.elasticsearch.search.aggregations.support.AggregationContext;
+import org.elasticsearch.search.lookup.SourceLookup;
+import org.elasticsearch.xpack.core.ml.job.config.CategorizationAnalyzerConfig;
+import org.elasticsearch.xpack.ml.job.categorization.CategorizationAnalyzer;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+public class CategorizeTextAggregator extends DeferableBucketAggregator {
+
+    private final TermsAggregator.BucketCountThresholds bucketCountThresholds;
+    private final SourceLookup sourceLookup;
+    private final BigArrays bigArrays;
+    private final MappedFieldType fieldType;
+    private final CategorizationAnalyzer analyzer;
+    private final String sourceFieldName;
+    private ObjectArray<CategorizationTokenTree> categorizers;
+    private final int maxChildren;
+    private final int maxDepth;
+    private final double similarityThreshold;
+    private final LongKeyedBucketOrds bucketOrds;
+
+    protected CategorizeTextAggregator(
+        String name,
+        AggregatorFactories factories,
+        AggregationContext context,
+        Aggregator parent,
+        String sourceFieldName,
+        MappedFieldType fieldType,
+        TermsAggregator.BucketCountThresholds bucketCountThresholds,
+        int maxChildren,
+        int maxDepth,
+        double similarityThreshold,
+        List<String> categorizationFilters,
+        Map<String, Object> metadata
+    ) throws IOException {
+        super(name, factories, context, parent, metadata);
+        this.sourceLookup = context.lookup().source();
+        this.sourceFieldName = sourceFieldName;
+        this.fieldType = fieldType;
+        CategorizationAnalyzerConfig categorizationAnalyzerConfig = CategorizationAnalyzerConfig.buildStandardCategorizationAnalyzer(
```

Comment:
As mentioned before, it shouldn't be hardcoded that we use the standard categorization analyzer.  Doing this is creating a problem for Chinese, Japanese and Korean users that will need to be fixed at some point, and I think it will be a lot simpler to fix in the first release than have to introduce complicated BWC code for mixed version clusters later on.